shopping-cart/app/shoppingcart/checkout/shipping/CheckoutShippingPageController.java

```diff
@@ -0,0 +1,140 @@
+                .thenComposeAsync(cart -> {
+                    if (shippingForm.hasErrors()) {
+                        return handleFormErrors(shippingForm, cart, userContext);
+                    } else {
+                        final String shippingMethodId = shippingForm.get().getShippingMethodId();
+                        return setShippingToCart(cart, shippingMethodId)
+                                .thenComposeAsync(updatedCart -> handleSuccessfulSetShipping(userContext), HttpExecution.defaultContext());
+                    }
+                }, HttpExecution.defaultContext());
+    }
+
+    protected CompletionStage<Cart> setShippingToCart(final Cart cart, final String shippingMethodId) {
+        final Reference<ShippingMethod> shippingMethodRef = ShippingMethod.referenceOfId(shippingMethodId);
+        final SetShippingMethod setShippingMethod = SetShippingMethod.of(shippingMethodRef);
+        return sphere().execute(CartUpdateCommand.of(cart, setShippingMethod));
+    }
+
+    protected CompletionStage<Result> handleSuccessfulSetShipping(final UserContext userContext) {
+        final Call call = reverseRouter().showCheckoutPaymentForm(userContext.locale().toLanguageTag());
+        return completedFuture(redirect(call));
+    }
+
+    protected CompletionStage<Result> handleFormErrors(final Form<CheckoutShippingFormData> shippingForm,
+                                                       final Cart cart, final UserContext userContext) {
+        return getShippingMethods(session())
+                .thenApplyAsync(shippingMethods -> {
+                    final ErrorsBean errors = new ErrorsBean(shippingForm);
+                    final CheckoutShippingPageContent pageContent = createPageContentWithShippingError(shippingForm, errors, shippingMethods);
+                    return badRequest(renderCheckoutShippingPage(cart, pageContent, userContext));
+                }, HttpExecution.defaultContext());
+    }
+
+    protected CompletionStage<Result> handleCtpError(final Throwable throwable, final Form<CheckoutShippingFormData> shippingForm,
+                                                     final Cart cart, final UserContext userContext) {
+        if (throwable.getCause() instanceof SphereException) {
+            return getShippingMethods(session())
+                    .thenApplyAsync(shippingMethods -> {
+                        Logger.error("Unknown CTP error", throwable.getCause());
+                        final ErrorsBean errors = new ErrorsBean("Something went wrong, please try again"); // TODO get from i18n
+                        final CheckoutShippingPageContent pageContent = createPageContentWithShippingError(shippingForm, errors, shippingMethods);
+                        return badRequest(renderCheckoutShippingPage(cart, pageContent, userContext));
+                    }, HttpExecution.defaultContext());
+        }
+        throw new RuntimeException(throwable);
```

Comment:
Ah I just realised this method is not even used, it was just a draft for the future handling of CTP errors. But I need your help with that because of `exceptionally`.